File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        vector<char> freq;
        for(int i = 0; i < s.size(); i++) {
            if (freq.size() != 0 && freq.back() == '(' && s[i] == ')') {
                freq.pop_back();
            }
            else if (freq.size() != 0 && freq.back() == '[' && s[i] == ']') {
                freq.pop_back();
            }
            else if (freq.size() != 0 && freq.back() == '{' && s[i] == '}') {
                freq.pop_back();
            }
            else {
                freq.push_back(s[i]);
            }    
            
        }
        if (freq.size() == 0) {
            return true;
        }
        return false;

    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []

        if len(s) % 2 != 0:
            return False
        

        for brackets in s:
            if brackets == ')' and stack and stack[-1] == '(':
                stack.pop()
            elif brackets == ']' and stack and stack[-1] == '[':
                stack.pop()
            elif brackets == '}' and stack and stack[-1] == '{':
                stack.pop()
            else:
                stack.append(brackets)
        return True if len(stack) == 0 else False

               